File: ui/src/lib/validate.ts
import { Ajv2020, type ErrorObject } from 'ajv/dist/2020';

import schema from '../../../schema/project.schema.json';

import { engineVersions } from './form.ts';
import { legalRelations } from './relations.ts';
import type { Node, NodeType, Project, ValidationError } from './types.ts';

const check = new Ajv2020({ allErrors: true, strict: false, verbose: true }).compile(schema);
const branches = schema.properties.nodes.items.oneOf.map(
  (variant) => variant.properties.type.const as NodeType,
);
const inBranch = /^#\/properties\/nodes\/items\/oneOf\/(\d+)\//;

// The same two passes as the CLI, in the same order and the same words, so a
// project the canvas calls sound is one the studio will save (ADR 0003).
export function validateProject(project: Project): ValidationError[] {
  const shape = schemaErrors(project);
  return shape.length > 0 ? shape : semanticErrors(project);
}

export function errorLine(error: ValidationError): string {
  const path = error.path === '' ? 'project' : error.path;
  if (error.nodeId !== undefined && error.nodeId !== '') {
    return `${path} (node ${error.nodeId}): ${error.message}`;
  }
  if (error.edgeId !== undefined && error.edgeId !== '') {
    return `${path} (edge ${error.edgeId}): ${error.message}`;
  }
  return `${path}: ${error.message}`;
}

function schemaErrors(project: Project): ValidationError[] {
  if (check(project)) {
    return [];
  }
  const out: ValidationError[] = [];
  for (const error of check.errors ?? []) {
    const at = segments(error.instancePath);
    if (error.keyword === 'oneOf' && at.length === 2 && at[0] === 'nodes') {
      const mistyped = mistypedNode(project, at);
      if (mistyped !== undefined) {
        out.push(mistyped);
      }
      continue;
    }
    // Six of the seven branches failed on the type const alone, and only the branch
    // the user meant has anything to say about the node.
    const branch = inBranch.exec(error.schemaPath)?.[1];
    if (branch !== undefined && Number(branch) !== branchFor(project, at)) {
      continue;
    }
    if (error.keyword === 'propertyNames') {
      continue;
    }
    out.push(entry(project, at, message(error)));
  }
  return out;
}

// A node is a oneOf over the seven types, discriminated by a const on one property,
// so the branch the user meant has already said what is wrong with the node.
function mistypedNode(project: Project, at: string[]): ValidationError | undefined {
  const node = project.nodes[Number(at[1])] as Node | undefined;
  const given = String(node?.type ?? '');
  if ((branches as string[]).includes(given)) {
    return undefined;
  }
  if (given === '') {
    return entry(project, at, "'oneOf' failed, none matched");
  }
  return {
    path: `${at.join('.')}.type`,
    nodeId: node?.id ?? '',
    message: `unknown node type '${given}', use one of ${branches.join(', ')}`,
  };
}

function branchFor(project: Project, at: string[]): number {
  const type = project.nodes[Number(at[1])]?.type;
  return branches.findIndex((candidate) => candidate === type);
}

function entry(project: Project, at: string[], text: string): ValidationError {
  const error: ValidationError = { path: at.join('.'), message: text };
  if (at.length < 2) {
    return error;
  }
  const index = Number(at[1]);
  if (at[0] === 'nodes') {
    error.nodeId = project.nodes[index]?.id ?? '';
  }
  if (at[0] === 'edges') {
    error.edgeId = project.edges[index]?.id ?? '';
  }
  return error;
}

function segments(instancePath: string): string[] {
  return instancePath === '' ? [] : instancePath.slice(1).split('/');
}

function message(error: ErrorObject): string {
  const params = error.params as Record<string, unknown>;
  const data: unknown = error.data;
  switch (error.keyword) {
    case 'required':
      return `missing property ${quote(String(params.missingProperty))}`;
    case 'additionalProperties':
      return `additional properties ${quote(String(params.additionalProperty))} not allowed`;
    case 'type':
      return `got ${jsonType(data)}, want ${String(params.type)}`;
    case 'enum':
      return oneOfMessage(params.allowedValues);
    case 'const':
      return `value must be ${display(params.allowedValue)}`;
    case 'pattern':
      return `${quote(String(data))} does not match pattern ${quote(String(params.pattern))}`;
    case 'minLength':
      return `minLength: got ${String(data).length}, want ${String(params.limit)}`;
    case 'maxLength':
      return `maxLength: got ${String(data).length}, want ${String(params.limit)}`;
    case 'minimum':
      return `minimum: got ${String(data)}, want ${String(params.limit)}`;
    case 'maximum':
      return `maximum: got ${String(data)}, want ${String(params.limit)}`;
    case 'minItems':
      return `minItems: got ${count(data)}, want ${String(params.limit)}`;
    case 'maxItems':
      return `maxItems: got ${count(data)}, want ${String(params.limit)}`;
    default:
      return error.message ?? 'validation failed';
  }
}

function oneOfMessage(allowed: unknown): string {
  const values = Array.isArray(allowed) ? (allowed as unknown[]) : [];
  if (values.length === 1) {
    return `value must be ${display(values[0])}`;
  }
  return `value must be one of ${values.map(display).join(', ')}`;
}

function count(data: unknown): number {
  return Array.isArray(data) ? data.length : 0;
}

function jsonType(data: unknown): string {
  if (data === null) {
    return 'null';
  }
  if (Array.isArray(data)) {
    return 'array';
  }
  return typeof data;
}

function display(value: unknown): string {
  return typeof value === 'string' ? quote(value) : String(value);
}

function quote(value: string): string {
  return `'${value.replaceAll("'", "\\'")}'`;
}

function semanticErrors(project: Project): ValidationError[] {
  const out: ValidationError[] = [];
  const byId = new Map<string, Node>();
  const names = new Set<string>();

  project.nodes.forEach((node, index) => {
    const path = `nodes.${index}`;
    if (byId.has(node.id)) {
      out.push({ path, nodeId: node.id, message: `duplicate node id '${node.id}'` });
    }
    if (names.has(node.name)) {
      out.push({ path, nodeId: node.id, message: `duplicate node name '${node.name}'` });
    }
    const unsupported = engineVersionError(project, node, index);
    if (unsupported !== undefined) {
      out.push(unsupported);
    }
    byId.set(node.id, node);
    names.add(node.name);
  });

  let gateways = 0;
  for (const node of project.nodes) {
    if (node.type !== 'gateway') {
      continue;
    }
    gateways += 1;
    if (gateways > 1) {
      out.push({ path: 'nodes', nodeId: node.id, message: 'a project can have at most one gateway' });
    }
  }

  const seen = new Set<string>();
  const ids = new Set<string>();
  project.edges.forEach((edge, index) => {
    const path = `edges.${index}`;
    if (ids.has(edge.id)) {
      out.push({ path, edgeId: edge.id, message: `duplicate edge id '${edge.id}'` });
    }
    ids.add(edge.id);

    const from = byId.get(edge.from);
    const to = byId.get(edge.to);
    if (from === undefined) {
      out.push({ path, edgeId: edge.id, message: `edge refers to missing node '${edge.from}'` });
    }
    if (to === undefined) {
      out.push({ path, edgeId: edge.id, message: `edge refers to missing node '${edge.to}'` });
    }
    if (from === undefined || to === undefined) {
      return;
    }
    if (from.id === to.id) {
      out.push({ path, edgeId: edge.id, message: 'an edge cannot connect a node to itself' });
      return;
    }
    if (!legalRelations(from.type, to.type).includes(edge.relation)) {
      out.push({
        path,
        edgeId: edge.id,
        message: `a ${from.type} cannot have a '${edge.relation}' edge to a ${to.type}`,
      });
    }
    const key = `${edge.from}|${edge.to}|${edge.relation}`;
    if (seen.has(key)) {
      out.push({
        path,
        edgeId: edge.id,
        message: `duplicate '${edge.relation}' edge from '${from.name}' to '${to.name}'`,
      });
    }
    seen.add(key);
  });

  return out;
}

// A provider with no engine table has no opinion yet, so nothing is checked for it.
function engineVersionError(
  project: Project,
  node: Node,
  index: number,
): ValidationError | undefined {
  if (node.type !== 'database') {
    return undefined;
  }
  const version = node.properties?.version;
  if (typeof version !== 'string' || version === '') {
    return undefined;
  }
  const held = node.properties?.engine;
  const engine = typeof held === 'string' && held !== '' ? held : 'postgres';
  const versions = engineVersions(project.provider, engine);
  if (versions === undefined || versions.includes(version)) {
    return undefined;
  }
  return {
    path: `nodes.${index}.properties.version`,
    nodeId: node.id,
    message: `database '${node.name}': ${engine} version '${version}' is not supported on ${project.provider} (use one of ${versions.join(', ')})`,
  };
}

File: internal/ir/validate.go
package ir

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"slices"
	"strconv"
	"strings"
	"sync"

	"github.com/santhosh-tekuri/jsonschema/v6"
	"github.com/santhosh-tekuri/jsonschema/v6/kind"
	"golang.org/x/text/language"
	"golang.org/x/text/message"
)

const projectSchemaURL = "https://togen.dev/schema/project.schema.json"

type ValidationError struct {
	Path    string `json:"path"`
	NodeID  string `json:"nodeId,omitempty"`
	EdgeID  string `json:"edgeId,omitempty"`
	Message string `json:"message"`
}

func (e ValidationError) String() string {
	path := e.Path
	if path == "" {
		path = "project"
	}
	switch {
	case e.NodeID != "":
		return fmt.Sprintf("%s (node %s): %s", path, e.NodeID, e.Message)
	case e.EdgeID != "":
		return fmt.Sprintf("%s (edge %s): %s", path, e.EdgeID, e.Message)
	}
	return fmt.Sprintf("%s: %s", path, e.Message)
}

type Errors []ValidationError

func (e Errors) Error() string {
	lines := make([]string, len(e))
	for i, err := range e {
		lines[i] = err.String()
	}
	return strings.Join(lines, "\n")
}

func ValidateProject(raw []byte) (*Project, Errors) {
	sch, err := projectSchema()
	if err != nil {
		return nil, Errors{{Message: err.Error()}}
	}
	instance, err := jsonschema.UnmarshalJSON(bytes.NewReader(raw))
	if err != nil {
		return nil, Errors{{Message: fmt.Sprintf("project file is not valid JSON: %v", err)}}
	}
	if err := sch.Validate(instance); err != nil {
		var invalid *jsonschema.ValidationError
		if !errors.As(err, &invalid) {
			return nil, Errors{{Message: err.Error()}}
		}
		return nil, schemaErrors(invalid, instance)
	}
	var p Project
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, Errors{{Message: err.Error()}}
	}
	if errs := semanticErrors(&p); len(errs) > 0 {
		return nil, errs
	}
	out, err := ApplyDefaults(&p)
	if err != nil {
		return nil, Errors{{Message: err.Error()}}
	}
	return out, nil
}

var (
	schemaOnce     sync.Once
	compiledSchema *jsonschema.Schema
	schemaErr      error
	printer        = message.NewPrinter(language.English)
)

func projectSchema() (*jsonschema.Schema, error) {
	schemaOnce.Do(func() {
		var doc any
		doc, schemaErr = jsonschema.UnmarshalJSON(bytes.NewReader(projectSchemaJSON))
		if schemaErr != nil {
			return
		}
		c := jsonschema.NewCompiler()
		if schemaErr = c.AddResource(projectSchemaURL, doc); schemaErr != nil {
			return
		}
		compiledSchema, schemaErr = c.Compile(projectSchemaURL)
	})
	return compiledSchema, schemaErr
}

func schemaErrors(root *jsonschema.ValidationError, instance any) Errors {
	var out Errors
	var walk func(e *jsonschema.ValidationError, loc []string)
	walk = func(e *jsonschema.ValidationError, loc []string) {
		// v6.0.3 aliases a shared buffer into a propertyNames error's location, and the
		// pattern failure beneath it carries none, so both keep the enclosing object's.
		if _, ok := e.ErrorKind.(*kind.PropertyNames); !ok && len(e.InstanceLocation) > 0 {
			loc = e.InstanceLocation
		}
		causes := e.Causes
		if _, ok := e.ErrorKind.(*kind.OneOf); ok {
			causes = discriminated(e, loc)
			if len(causes) == 0 {
				if unknown, ok := unknownNodeType(loc, instance); ok {
					out = append(out, unknown)
					return
				}
			}
		}
		if len(causes) == 0 {
			out = append(out, schemaError(loc, e.ErrorKind.LocalizedString(printer), instance))
			return
		}
		for _, cause := range causes {
			walk(cause, loc)
		}
	}
	walk(root, root.InstanceLocation)
	return out
}

// A node is a oneOf over the seven types, discriminated by a const on one property.
// Only the branch the user meant is worth reporting; the other six failed on the const.
func discriminated(e *jsonschema.ValidationError, loc []string) []*jsonschema.ValidationError {
	var kept []*jsonschema.ValidationError
	for _, cause := range e.Causes {
		if !failsDiscriminator(cause, len(loc)+1) {
			kept = append(kept, cause)
		}
	}
	if len(kept) == 1 {
		return kept
	}
	return nil
}

func failsDiscriminator(e *jsonschema.ValidationError, depth int) bool {
	if _, ok := e.ErrorKind.(*kind.Const); ok && len(e.InstanceLocation) == depth {
		return true
	}
	for _, cause := range e.Causes {
		if failsDiscriminator(cause, depth) {
			return true
		}
	}
	return false
}

// Every branch failed its type const, so the raw oneOf wording says nothing useful.
func unknownNodeType(loc []string, instance any) (ValidationError, bool) {
	if len(loc) != 2 || loc[0] != "nodes" {
		return ValidationError{}, false
	}
	node, ok := itemAt(instance, loc[0], loc[1])
	if !ok {
		return ValidationError{}, false
	}
	given, ok := node["type"].(string)
	if !ok || given == "" || slices.Contains(NodeTypes, NodeType(given)) {
		return ValidationError{}, false
	}
	names := make([]string, len(NodeTypes))
	for i, t := range NodeTypes {
		names[i] = string(t)
	}
	id, _ := node["id"].(string)
	return ValidationError{
		Path:    strings.Join(loc, ".") + ".type",
		NodeID:  id,
		Message: fmt.Sprintf("unknown node type '%s', use one of %s", given, strings.Join(names, ", ")),
	}, true
}

func schemaError(loc []string, message string, instance any) ValidationError {
	out := ValidationError{Path: strings.Join(loc, "."), Message: message}
	if len(loc) >= 2 {
		id := idAt(instance, loc[0], loc[1])
		switch loc[0] {
		case "nodes":
			out.NodeID = id
		case "edges":
			out.EdgeID = id
		}
	}
	return out
}

func idAt(instance any, collection, index string) string {
	item, ok := itemAt(instance, collection, index)
	if !ok {
		return ""
	}
	id, _ := item["id"].(string)
	return id
}

func itemAt(instance any, collection, index string) (map[string]any, bool) {
	doc, ok := instance.(map[string]any)
	if !ok {
		return nil, false
	}
	items, ok := doc[collection].([]any)
	if !ok {
		return nil, false
	}
	i, err := strconv.Atoi(index)
	if err != nil || i < 0 || i >= len(items) {
		return nil, false
	}
	item, ok := items[i].(map[string]any)
	return item, ok
}

// A provider with no engine table has no opinion yet, so nothing is checked for it.
func engineVersionError(p *Project, n *Node, i int) (ValidationError, bool) {
	if n.Type != NodeDatabase {
		return ValidationError{}, false
	}
	props, err := NodeProps[DatabaseProps](*n)
	if err != nil || props.Version == "" {
		return ValidationError{}, false
	}
	engine := props.Engine
	if engine == "" {
		engine = EnginePostgres
	}
	table, ok := Engines[p.Provider]
	if !ok {
		return ValidationError{}, false
	}
	info, ok := table[engine]
	if !ok || slices.Contains(info.Versions, props.Version) {
		return ValidationError{}, false
	}
	return ValidationError{
		Path:   fmt.Sprintf("nodes.%d.properties.version", i),
		NodeID: n.ID,
		Message: fmt.Sprintf("database '%s': %s version '%s' is not supported on %s (use one of %s)",
			n.Name, engine, props.Version, p.Provider, strings.Join(info.Versions, ", ")),
	}, true
}

func semanticErrors(p *Project) Errors {
	var errs Errors
	byID := make(map[string]*Node, len(p.Nodes))
	names := make(map[string]bool, len(p.Nodes))
	gateways := 0

	for i := range p.Nodes {
		n := &p.Nodes[i]
		path := fmt.Sprintf("nodes.%d", i)
		if _, ok := byID[n.ID]; ok {
			errs = append(errs, ValidationError{Path: path, NodeID: n.ID, Message: fmt.Sprintf("duplicate node id '%s'", n.ID)})
		}
		if names[n.Name] {
			errs = append(errs, ValidationError{Path: path, NodeID: n.ID, Message: fmt.Sprintf("duplicate node name '%s'", n.Name)})
		}
		if err, bad := engineVersionError(p, n, i); bad {
			errs = append(errs, err)
		}
		byID[n.ID] = n
		names[n.Name] = true
	}

	for i := range p.Nodes {
		if p.Nodes[i].Type != NodeGateway {
			continue
		}
		gateways++
		if gateways > 1 {
			errs = append(errs, ValidationError{Path: "nodes", NodeID: p.Nodes[i].ID, Message: "a project can have at most one gateway"})
		}
	}

	seen := make(map[string]bool, len(p.Edges))
	edgeIDs := make(map[string]bool, len(p.Edges))
	for i, e := range p.Edges {
		path := fmt.Sprintf("edges.%d", i)
		if edgeIDs[e.ID] {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: fmt.Sprintf("duplicate edge id '%s'", e.ID)})
		}
		edgeIDs[e.ID] = true

		from, fromOK := byID[e.From]
		to, toOK := byID[e.To]
		if !fromOK {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: fmt.Sprintf("edge refers to missing node '%s'", e.From)})
		}
		if !toOK {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: fmt.Sprintf("edge refers to missing node '%s'", e.To)})
		}
		if !fromOK || !toOK {
			continue
		}
		if from.ID == to.ID {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: "an edge cannot connect a node to itself"})
			continue
		}
		if !slices.Contains(LegalRelations(from.Type, to.Type), e.Relation) {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: fmt.Sprintf("a %s cannot have a '%s' edge to a %s", from.Type, e.Relation, to.Type)})
		}
		key := e.From + "|" + e.To + "|" + string(e.Relation)
		if seen[key] {
			errs = append(errs, ValidationError{Path: path, EdgeID: e.ID, Message: fmt.Sprintf("duplicate '%s' edge from '%s' to '%s'", e.Relation, from.Name, to.Name)})
		}
		seen[key] = true
	}

	return errs
}
